soc/arm/microchip_mec/common/soc_dma.c

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2021 Microchip Technology Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <sys/sys_io.h>
+#include <sys/__assert.h>
+#include <soc.h>
+
+/* #define SOC_DMA_DEBUG_ENABLE 1 */
+
+int soc_dma_init(void)
+{
+	mec_pcr_periph_slp_ctrl(PCR_DMA, 0);
+
+	struct dma_regs *regs = (struct dma_regs *)(DMA_BASE);
+
+	regs->ACTRST = MCHP_DMAM_CTRL_SOFT_RESET | MCHP_DMAM_CTRL_ENABLE;
+
+	uint32_t cnt = 256U;
+
+	while (cnt--) {
+		if (regs->ACTRST == 0) {
+			regs->ACTRST = MCHP_DMAM_CTRL_ENABLE;
+			return 0;
+		}
+	}
+
+	return -ETIMEDOUT;
+}
+
+int soc_dma_chan_init(uint8_t chan)
+{
+	if (chan >= MCHP_DMA_CHANNELS) {
+		return -EINVAL;
+	}
+
+	struct dma_chan_regs *regs =
+		(struct dma_chan_regs *)DMA_CHAN_BASE(chan);
+
+	if (regs->ACTV) {
+		regs->CTRL |= MCHP_DMA_C_XFER_ABORT;
+
+		uint32_t cnt = 256U;
+
+		while (cnt--) {
+			if (!(regs->CTRL & MCHP_DMA_C_BUSY_STS)) {
+				break;
+			}
+		}
+
+		if (regs->CTRL & MCHP_DMA_C_BUSY_STS) {
+			return -ETIMEDOUT;
+		}
+
+		regs->ACTV = 0U;
+	}
+
+	regs->CTRL = 0U;
+	regs->MSTART = 0U;
+	regs->MEND = 0U;
+	regs->IEN = 0U;
+	regs->ISTS = 0xFFFFFFFFul;
+
+	return 0;
+}
+
+int soc_dma_chan_cfg_mem(uint8_t chan, uintptr_t mbase, uint32_t sz)
+{
+	if (chan >= MCHP_DMA_CHANNELS) {
+		return -EINVAL;
+	}
+
+	struct dma_chan_regs *regs =
+		(struct dma_chan_regs *)DMA_CHAN_BASE(chan);
+
+	regs->MSTART = mbase;
+	regs->MEND = mbase + sz;
+
+	return 0;
+}
+
+int soc_dma_chan_cfg_device(uint8_t chan, uint32_t dev_dma_id,
+			    uintptr_t dev_addr)
+{
+	if (chan >= MCHP_DMA_CHANNELS) {
+		return -EINVAL;
+	}
+
+	struct dma_chan_regs *regs =
+		(struct dma_chan_regs *)DMA_CHAN_BASE(chan);
+
+	uint32_t ctrl = regs->CTRL & ~(MCHP_DMA_C_DEV_NUM_MASK);
```

Comment:
Drop unnecessary () again. Please fix globally

Reply:
It's good practice to use `~(...)` to ensure that the result of the inner expression is processed (and not just the first item of the inner expression).

Reply:
I was taught that expressions in #defines should have brackets:

#define MCHP_DMA_C_DEV_NUM_MASK (1 + 2)

to avoid this problem